dashboard/lib/policies/lti.rb

```diff
@@ -12,4 +15,39 @@ module AccessTokenScopes
   JWT_CLIENT_ASSERTION_TYPE = 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer'.freeze
   JWT_ISSUER = CDO.studio_url('', CDO.default_scheme).freeze
   MEMBERSHIP_CONTAINER_CONTENT_TYPE = 'application/vnd.ims.lti-nrps.v2.membershipcontainer+json'.freeze
+  TEACHER_ROLES = Set.new(
+    [
+      'http://purl.imsglobal.org/vocab/lis/v2/institution/person#Administrator',
+      'http://purl.imsglobal.org/vocab/lis/v2/institution/person#Instructor',
+      'http://purl.imsglobal.org/vocab/lis/v2/membership#Administrator',
+      'http://purl.imsglobal.org/vocab/lis/v2/membership#Instructor',
+      'http://purl.imsglobal.org/vocab/lis/v2/system/person#Administrator',
+    ]
+).freeze
+  LTI_ROLES_KEY = 'https://purl.imsglobal.org/spec/lti/claim/roles'.freeze
+  LTI_CUSTOM_CLAIMS = "https://purl.imsglobal.org/spec/lti/claim/custom".freeze
+
+  def self.get_account_type(id_token)
+    id_token[LTI_ROLES_KEY].each do |role|
+      return User::TYPE_TEACHER if TEACHER_ROLES.include? role
+    end
+    return User::TYPE_STUDENT
+  end
+
+  def self.generate_auth_id(id_token)
+    "#{id_token[:iss]}|#{id_token[:aud]}|#{id_token[:sub]}"
```

Comment:
Do we need a validation in the auth options model to ensure that for LTI auth options, the ID conforms to this pattern?

Reply:
Would the validation essentially check for two `|` to present in the string?

Reply:
Yes and that there are some characters in between. The intention is mostly to guard against something in the input being nil, and ending up with an id like `||`, so it might actually fit better inside this `generate_auth_id` function; we could raise an argument error if iss/aug/sub are empty/nil. A validation would have the extra benefit of making sure nobody creates an LTI auth option with an ID in a different format. Not saying our current team would, but say years from now this generate_auth_id function would be the only indicator of why this ID is the way it is.